src/dir_set.ml

```diff
@@ -0,0 +1,180 @@
+    | _, res -> Some res)
+
+let merge_nontrivial a b ~f_one ~f_set =
+  let default = f_one a.default b.default in
+  create
+    ~here:(f_one a.here b.here)
+    ~default
+    ~exceptions:(merge_exceptions a b ~default ~f:f_set)
+[@@inline always]
+
+let rec union x y =
+  match x, y with
+  | Empty, v | v, Empty -> v
+  | Universal, _ | _, Universal -> Universal
+  | Nontrivial x, Nontrivial y ->
+    merge_nontrivial x y ~f_one:(||) ~f_set:union
+
+let rec inter x y =
+  match x, y with
+  | Universal, v | v, Universal -> v
+  | Empty, _ | _, Empty -> Empty
+  | Nontrivial x, Nontrivial y ->
+    merge_nontrivial x y ~f_one:(&&) ~f_set:inter
+
+let rec negate x =
+  match x with
+  | Universal -> Empty
+  | Empty -> Universal
+  | Nontrivial { here; default; exceptions } ->
+    Nontrivial { here = not here
+               ; default = not default
+               ; exceptions = String.Map.map exceptions ~f:negate
+               }
+
+let rec diff x y =
+  match x with
+  | Empty -> Empty
+  | Universal -> negate y
+  | Nontrivial nx ->
+    match y with
+    | Empty -> x
+    | Universal -> Empty
+    | Nontrivial ny ->
+      merge_nontrivial nx ny
+        ~f_one:(fun a b -> a && not b)
+        ~f_set:diff
+
+let rec mem t dir =
+  match t with
+  | Empty -> false
+  | Universal -> true
+  | Nontrivial { here; default; exceptions } ->
+    match dir with
+    | [] -> here
+    | child :: rest ->
+      match String.Map.find exceptions child with
+      | None -> default
+      | Some t -> mem t rest
+
+let mem t dir = mem t (Path.Build.explode dir)
+
+let descend t child =
+  match t with
+  | Empty -> Empty
+  | Universal -> Universal
+  | Nontrivial { here = _; default; exceptions } ->
+    match String.Map.find exceptions child with
+    | None -> trivial default
+    | Some t -> t
+
+let union_all = List.fold_left ~init:empty ~f:union
+let inter_all = List.fold_left ~init:empty ~f:inter
+
+let of_subtree_gen =
+  let rec loop subtree = function
+    | [] -> subtree
+    | component :: rest ->
+      create
+        ~here:false
+        ~default:false
+        ~exceptions:(String.Map.singleton component (loop subtree rest))
+  in
+  fun subtree path -> loop subtree (Path.Build.explode path)
+
+let just_the_root =
+  Nontrivial
+    { here = true
+    ; default = false
+    ; exceptions = String.Map.empty
+    }
+
+let subtree p = of_subtree_gen universal p
+let singleton p = of_subtree_gen just_the_root p
+
+let is_subset =
+  let not_subset () = raise_notrace Exit in
+  let rec loop x y =
+    match x, y with
+    | _, Universal | Empty, _ -> Empty
+    | Universal, _ | _, Empty -> not_subset ()
+    | Nontrivial x, Nontrivial y ->
+      if (x.here    && not y.here   ) ||
+         (x.default && not y.default) then
+        not_subset ();
+      ignore
+        (merge_exceptions x y ~default:false ~f:loop : t String.Map.t);
+      Empty
+  in
+  fun x ~of_ ->
+    match loop x of_ with
+    | (_ : t) -> true
```

Comment:
Why do we return `t` here and not `unit` for example?

Reply:
It was just so that I could write `merge... ~f:loop` which wants `f`  to return a `t`.